pkg/discovery/util/pod_util.go

```diff
@@ -276,3 +281,127 @@ func GetPodGrandInfo(kclient *client.Clientset, pod *api.Pod) (string, string, e
+			glog.Errorf("Pod %s: %s", podName, pe.message)
+		} else {
+			glog.V(2).Infof("Pod %s: %s", podName, pe.message)
+		}
+	}
+	return err
+}
+
+func checkPodNode(kubeClient *client.Clientset, namespace, podName, nodeName string) (bool, error) {
+	pod, err := kubeClient.CoreV1().Pods(namespace).Get(podName, metav1.GetOptions{})
+	if err != nil {
+		return true, err
+	}
+	if pod.Status.Phase == api.PodRunning && PodIsReady(pod) {
+		if len(nodeName) > 0 {
+			if !strings.EqualFold(pod.Spec.NodeName, nodeName) {
+				return false, fmt.Errorf("pod %s is running on an unexpected node %v", podName, pod.Spec.NodeName)
+			}
+		}
+		return false, nil
+	}
+	warnings := getPodWarnings(pod, podName)
+	return true, fmt.Errorf("%s", strings.Join(warnings, ", "))
+}
+
+// getPodWarnings gets a list of short warning messages that belong to the given pod
+// These messages will be concatenated and sent to the UI
+// The warning messages are logged in the following order:
+//   - Pod states and reasons
+//   - Pod conditions and reasons (false conditions only)
+//   - Container states and reasons (abnormal states only)
+// https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-and-container-status
+func getPodWarnings(pod *api.Pod, podName string) (warnings []string) {
+	warnings = []string{}
+	// pod statuses
+	podState := ""
+	if pod.DeletionTimestamp != nil {
+		podState = fmt.Sprintf("pod %s is being deleted", podName)
+	} else {
+		podState = fmt.Sprintf("pod %s is in %s state", podName, pod.Status.Phase)
+	}
+	if pod.Status.Reason != "" {
+		podState = podState + " due to " + pod.Status.Reason
+	}
+	warnings = append(warnings, podState)
+	// pod conditions
+	visited := make(map[string]bool, 0)
+	for _, condition := range pod.Status.Conditions {
+		// skip true conditions
+		if condition.Status == api.ConditionTrue ||
+			visited[condition.Reason] {
+			continue
+		}
+		visited[condition.Reason] = true
+		warnings = append(warnings, condition.Reason)
+	}
+	// container statuses
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		name := containerStatus.Name
+		state := containerStatus.State
+		if state.Waiting != nil {
+			warnings = append(warnings,
+				fmt.Sprintf("container %s is waiting due to %s",
+					name, state.Waiting.Reason))
+		}
+		if state.Terminated != nil {
+			warnings = append(warnings,
+				fmt.Sprintf("container %s has terminated due to %s",
+					name, state.Terminated.Reason))
+		}
+	}
+	return
+}
+
+// getPodEvents gets a list of unique events that belong to the given pod
+// These events can be very long, and will only be written to the log file
+func getPodEvents(kubeClient *client.Clientset, namespace, name string) (podEvents []podEvent) {
+	podEvents = []podEvent{}
+	// Get the pod
+	pod, err := kubeClient.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return
+	}
+	// Get all events
+	events, err := kubeClient.CoreV1().Events(namespace).List(metav1.ListOptions{})
```

Comment:
Can we use the ListOptions to select only the events relevant? Some example using ListOptions is as below:
podList, err := podClient.List(meta_v1.ListOptions{
		FieldSelector: "status.phase=" + string(phase) + ",metadata.uid=" + uid,
	})

Reply:
Resolved